Training24/Test.c

```diff
@@ -1,19 +1,172 @@
+   const char* phrases[] = { "Race!Car", "hello", "D", "Don't nod", "A man, a Plan, a canal, Panama",
+      " ","22 / 02 / 2022", "12345654321" };
+   for (int i = 0; i < sizeof (phrases) / sizeof (phrases[0]); i++) {
+      int isPhrasePalindrome = CheckPhrasePalindrome (phrases[i]);
+      PhraseTableRow (phrases[i], isPhrasePalindrome ? GREEN "It's a Palindrome" RESET : RED "Not a Palindrome" RESET);
+   }
+   printf ("\n\n");
+}
+
+void TestReverseNumbers () {
+   IntegerTableHeader ();
+   int testCases[] = { 212, -121, 123421, -9876789, 2147447412, 2147483647, -2147483648LL, 999999999 };
+   for (int i = 0; i < sizeof (testCases) / sizeof (testCases[0]); i++) {
+      int number = testCases[i];
+      if (number < 0) IntegerTableRow (number, "Invalid(Negative number)", false);
+      else {
+         int reversedNumber;
+         bool overflow = ReverseAndCheckOverflow (number, &reversedNumber);
+         char outputStr[256];
+         if (overflow) IntegerTableRow (number, "Integer Overflow", false);
+         else {
+            sprintf (outputStr, "%d", reversedNumber);
+            bool isPalindrome = (number == reversedNumber);
+            IntegerTableRow (number, outputStr, isPalindrome);
+         }
+      }
+   }
+}
+
+void clearInputBuffer () {
+   while (getchar () != '\n');
+}
+
+void TestUserPhrase () {
+   char phrase[256];
+   printf ("\nEnter a phrase to test: ");
+   fgets (phrase, sizeof (phrase), stdin);
+   phrase[strcspn (phrase, "\n")] = 0;
+   printf ("Result: %s\n", phrase[0] == '\0' ? GREEN "It's a Palindrome" RESET :
+      CheckPhrasePalindrome (phrase) ? GREEN "It's a Palindrome" RESET : RED "Not a Palindrome" RESET);
+}
+
+void TestUserInteger () {
+   long inputNumber;
+   char userInput[256];
+   while (1) {
+      printf ("\nEnter a positive integer to test: ");
+      if (!fgets (userInput, sizeof (userInput), stdin) || userInput[0] == '\n' || userInput == NULL) {
+         printf ("Result: %s\n", YELLOW "Invalid input (Empty string)" RESET);
+         continue;
+      }
+      char* endPtr;
+      inputNumber = strtol (userInput, &endPtr, 10);
+      if (endPtr == userInput || *endPtr != '\n' || inputNumber < 0) {
+         printf ("Result: %s\n", (endPtr == userInput || *endPtr != '\n') ?
+            YELLOW "Invalid (Non-integer input)" RESET :
+            YELLOW "Invalid (Negative number is not a palindrome)" RESET);
+         continue;
+      }
+      else {
+         int validNumber = (int)inputNumber; // Cast to int after valid input
+         int reversedNumber;
+         bool overflow = ReverseAndCheckOverflow (validNumber, &reversedNumber);
+         if (overflow) printf ("Result: %s\n", YELLOW "Invalid (Out of range)" RESET);
+         else {
+            bool isPalindrome = IsNumPalindrome (validNumber);
+            printf ("Output: %d\nResult: %s\n", reversedNumber,
+               isPalindrome ? GREEN "It's a Palindrome" RESET : RED "Not a Palindrome" RESET);
+            break;
```

Comment:
You can continue even if it is a valid input. Otherwise, the user would have to repeat making choices for each input. Have a separate key to exit the loop and ask the user to choose again if the user wishes to.